components/sync15/src/sync_multiple.rs

```diff
@@ -102,160 +130,301 @@ pub fn sync_multiple(
-    mem_cached_state: &mut MemoryCachedState,
-    storage_init: &Sync15StorageClientInit,
-    root_sync_key: &KeyBundle,
-    interruptee: &impl Interruptee,
-    sync_result: &mut SyncResult,
-) -> result::Result<(), Error> {
-    if interruptee.was_interrupted() {
-        sync_result.service_status = ServiceStatus::Interrupted;
-        return Ok(());
-    }
+/// This is essentially a bag of information that the sync manager knows, but
+/// otherwise we won't. It should probably be rethought if it gains many more
+/// fields.
+#[derive(Debug, Default)]
+pub struct SyncRequestInfo<'a> {
+    pub engines_to_state_change: Option<&'a HashMap<String, bool>>,
+    pub is_user_action: bool,
+}
 
-    // We put None back into last_client_info now so if we fail entirely,
-    // reinitialize everything related to the client.
-    let client_info = match mem::replace(&mut mem_cached_state.last_client_info, None) {
-        Some(client_info) => {
-            // if our storage_init has changed it probably means the user has
-            // changed, courtesy of the 'kid' in the structure. Thus, we can't
-            // reuse the client or the memory cached state. We do keep the disk
-            // state as currently that's only the declined list.
-            if client_info.client_init != *storage_init {
-                log::info!("Discarding all state as the account might have changed");
-                *mem_cached_state = MemoryCachedState::default();
-                ClientInfo::new(storage_init)?
-            } else {
-                // we can reuse it (which should be the common path)
-                client_info
-            }
+// The sync multiple driver
+struct SyncMultipleDriver<'info, 'res, 'pgs, 'mcs> {
+    stores: &'info [&'info dyn Store],
+    storage_init: &'info Sync15StorageClientInit,
+    root_sync_key: &'info KeyBundle,
+    interruptee: &'info dyn Interruptee,
+    backoff: BackoffListener,
+    engines_to_state_change: Option<&'info HashMap<String, bool>>,
+    result: &'res mut SyncResult,
+    persisted_global_state: &'pgs mut Option<String>,
+    mem_cached_state: &'mcs mut MemoryCachedState,
+    ignore_soft_backoff: bool,
+    any_failed_engines: bool,
+}
+
+impl<'info, 'res, 'pgs, 'mcs> SyncMultipleDriver<'info, 'res, 'pgs, 'mcs> {
+    /// The actual worker for sync_multiple.
+    fn sync(mut self) -> result::Result<(), Error> {
+        log::info!("Loading/initializing persisted state");
+        let mut pgs = self.prepare_persisted_state();
+
+        log::info!("Preparing client info");
+        let client_info = self.prepare_client_info()?;
+
+        if self.was_interrupted() {
+            return Ok(());
         }
-        None => {
-            // We almost certainly have no other state here, but to be safe, we
-            // throw away any memory state we do have.
-            *mem_cached_state = MemoryCachedState::default();
-            ClientInfo::new(storage_init)?
+
+        if self.was_interrupted() {
```

Comment:
GitHub really made a hash of this diff! 😕 But I think the `self.was_interrupted` check is here twice.